src/metabase/public_settings/premium_features.clj

```diff
@@ -105,38 +106,42 @@
 
 (mu/defn ^:private fetch-token-status* :- TokenStatus
   "Fetch info about the validity of `token` from the MetaStore."
-  [token :- ValidToken]
+  [token :- :string]
   ;; attempt to query the metastore API about the status of this token. If the request doesn't complete in a
   ;; reasonable amount of time throw a timeout exception
-  (log/info (trs "Checking with the MetaStore to see whether {0} is valid..."
-                 ;; ValidToken will ensure the length of token is 64 chars long
-                 (str (subs token 0 4) "..." (subs token 60 64))))
-  (let [fut    (future
-                 (try (fetch-token-and-parse-body token token-check-url)
-                      (catch Exception e1
-                        (log/error e1 (trs "Error fetching token status from {0}:" token-check-url))
-                        ;; Try the fallback URL, which was the default URL prior to 45.2
-                        (try (fetch-token-and-parse-body token store-url)
-                             ;; if there was an error fetching the token from both the normal and fallback URLs, log the
-                             ;; first error and return a generic message about the token being invalid. This message
-                             ;; will get displayed in the Settings page in the admin panel so we do not want something
-                             ;; complicated
-                             (catch Exception e2
-                               (log/error e2 (trs "Error fetching token status from {0}:" store-url))
-                               (let [body (u/ignore-exceptions (some-> (ex-data e1) :body (json/parse-string keyword)))]
-                                 (or
-                                  body
-                                  {:valid         false
-                                   :status        (tru "Unable to validate token")
-                                   :error-details (.getMessage e1)})))))))
-        result (deref fut fetch-token-status-timeout-ms ::timed-out)]
-    (if (= result ::timed-out)
-      (do
-        (future-cancel fut)
-        {:valid         false
-         :status        (tru "Unable to validate token")
-         :error-details (tru "Token validation timed out.")})
-      result)))
+  (log/infof "Checking with the MetaStore to see whether token '%s' is valid..." (u.str/mask token))
+  (if-not (mc/validate ValidToken token)
```

Comment:
I think we can do this earlier in the call chain, how about checking this in the getter of `token-status`?

also, I wonder why it doesn't go through the setter of `premium-embedding-token`, we already have a schema check there.

Reply:
If I understood code correctly, then `:setter` is only called when you set something from inside Metabase' codebase. With env var, Metabase starts up and then when you call `(premium-embedding-token)`, it's going through `settings/get`, which eventually gets value from the environment. I agree it would be better if `register-setting!` or whoever is at fault should call `:setter`, but this looks like a much bigger (and not totally backward compatible) change?